__tests__/privateRpc.ts

```diff
@@ -0,0 +1,262 @@
+
+let con: DataSource;
+
+beforeAll(async () => {
+  con = await createOrGetConnection();
+});
+
+beforeEach(async () => {
+  await saveFixtures(con, Source, [
+    ...sourcesFixture,
+    {
+      id: 'collections',
+      name: 'Collections',
+      image: 'http://image.com/collections',
+      handle: 'collections',
+      type: SourceType.Machine,
+    },
+  ]);
+});
+
+const mockTransport = createRouterTransport(privateRpc, {
+  router: {
+    interceptors: [
+      (next) => {
+        return async (req) => {
+          if (
+            req.header.get('Authorization') ===
+            `Service ${process.env.SERVICE_SECRET}`
+          ) {
+            req.contextValues.set(baseRpcContext, {
+              service: true,
+            });
+          }
+
+          return next(req);
+        };
+      },
+    ],
+  },
+});
+const mockClient = createPromiseClient(PostService, mockTransport);
+const defaultClientAuthOptions: CallOptions = {
+  headers: {
+    Authorization: `Service ${process.env.SERVICE_SECRET}`,
+  },
+};
+
+describe('PostService', () => {
+  it('should return not found when not authorized', async () => {
+    baseRpcContext.defaultValue = {
+      service: false,
+    };
+
+    await expect(
+      mockClient.create({
+        url: 'http://example.com/service/1',
+        sourceId: 'a',
+      }),
+    ).rejects.toThrow(
+      new ConnectError('unauthenticated', Code.Unauthenticated),
+    );
+  });
+
+  it('should create post', async () => {
+    const result = await mockClient.create(
+      {
+        url: 'http://example.com/service/1',
+        sourceId: 'a',
+      },
+      defaultClientAuthOptions,
+    );
+
+    expect(result).toEqual({
+      postId: expect.any(String),
+      url: 'http://example.com/service/1',
+    });
+    const post = await con
+      .getRepository(ArticlePost)
+      .findOneBy({ id: result.postId });
+    expect(post).toBeTruthy();
+  });
+
+  it('should return duplicate entry', async () => {
+    const { postId } = await mockClient.create(
+      {
+        url: 'http://example.com/service/1',
+        sourceId: 'a',
+      },
+      defaultClientAuthOptions,
+    );
+    const post = await con.getRepository(ArticlePost).findOneBy({ id: postId });
+    expect(post).toBeTruthy();
+
+    const result = await mockClient.create(
+      {
+        url: 'http://example.com/service/1',
+        sourceId: 'a',
+      },
+      defaultClientAuthOptions,
+    );
+
+    expect(result).toEqual({
+      postId,
+      url: 'http://example.com/service/1',
+    });
+  });
+
+  it('should throw on invalid source', async () => {
+    await expect(
+      mockClient.create(
+        {
+          url: 'http://example.com/service/1',
+          sourceId: 'sourceDoestNotExist',
+        },
+        defaultClientAuthOptions,
+      ),
+    ).rejects.toThrow(new ConnectError('source not found', Code.NotFound));
+  });
+
+  it('should clean url', async () => {
```

Comment:
Not blocking but maybe add one for a excluded from cleaning url?
Just in case someone ever accidentally removes something there 🥲

Reply:
Yep, added here 4d37654e5498782b8fdf9e62b8901a7667747f43